File: android/app/src/main/kotlin/com/wavepaths/player/foreground/types/enums/ForegroundServiceMethod.kt
package com.wavepaths.player.foreground.types.enums

import android.util.Log
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
enum class ForegroundServiceMethod {
    @SerialName("init") INIT,
    @SerialName("start") START,
    @SerialName("startSessionEarly") START_SESSION_EARLY,
    @SerialName("broadcastUserAdvanceFromPrelude") BROADCAST_USER_ADVANCE_FROM_PRELUDE,
    @SerialName("resume") RESUME,
    @SerialName("pause") PAUSE,
    @SerialName("dispose") DISPOSE;

    companion object {
        private const val TAG = "ForegroundServiceMethod"
        fun fromString(value: String): ForegroundServiceMethod? {
            return try {
                Json.decodeFromString<ForegroundServiceMethod>("\"$value\"")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse enum from string: $value")
                null
            }
        }
    }
}

File: ios/Runner/Foreground/ForegroundServiceMethod.swift
import Foundation

enum ForegroundServiceMethod: String {
  case start = "start"
  case startSessionEarly = "startSessionEarly"
  case broadcastUserAdvanceFromPrelude = "broadcastUserAdvanceFromPrelude"
  case resume = "resume"
  case pause = "pause"
  case dispose = "dispose"

  func serialize() -> String? {
    let payload: [String: String] = ["type": rawValue]
    guard let data = try? JSONSerialization.data(withJSONObject: payload, options: []) else {
      return nil
    }
    return String(data: data, encoding: .utf8)
  }
}
